File: windows-native/PitlaneDanmaku.Windows/Services/TextSanitizer.cs
using System.Text;
using PitlaneDanmaku.Windows.Models;

namespace PitlaneDanmaku.Windows.Services;

public static class TextSanitizer
{
    public static ChatMessage? Normalize(ChatMessage message, AppSettings settings)
    {
        var userName = CleanText(message.UserName, settings.MaxNicknameLength, settings.MaxRepeatCharacters);
        var text = CleanText(message.Text, settings.MaxMessageLength, settings.MaxRepeatCharacters);

        if (string.IsNullOrWhiteSpace(userName))
        {
            userName = "匿名观众";
        }

        if (!message.IsSuperChat && string.IsNullOrWhiteSpace(text))
        {
            return null;
        }

        return message with
        {
            UserName = userName,
            Text = string.IsNullOrWhiteSpace(text) ? " " : text
        };
    }

    private static string CleanText(string value, int maxLength, int repeatLimit)
    {
        var builder = new StringBuilder();
        var previous = '\0';
        var repeatCount = 0;

        foreach (var raw in value)
        {
            var ch = char.IsControl(raw) || char.IsWhiteSpace(raw) ? ' ' : raw;
            if (ch == previous)
            {
                repeatCount++;
                if (repeatCount > repeatLimit)
                {
                    continue;
                }
            }
            else
            {
                previous = ch;
                repeatCount = 1;
            }

            if (ch == ' ' && builder.Length > 0 && builder[^1] == ' ')
            {
                continue;
            }

            builder.Append(ch);
            if (builder.Length >= maxLength)
            {
                break;
            }
        }

        return builder.ToString().Trim();
    }
}

File: macos-native/Sources/PitlaneDanmakuMac/TextSanitizer.swift
import Foundation

enum TextSanitizer {
    static func normalize(_ message: ChatMessage, settings: AppSettings) -> ChatMessage? {
        var normalized = message
        normalized.userName = cleanText(message.userName, maxLength: settings.maxNicknameLength, repeatLimit: settings.maxRepeatCharacters)
        normalized.text = cleanText(message.text, maxLength: settings.maxMessageLength, repeatLimit: settings.maxRepeatCharacters)

        if normalized.userName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            normalized.userName = "匿名观众"
        }

        if !normalized.isSuperChat && normalized.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return nil
        }

        if normalized.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            normalized.text = " "
        }

        return normalized
    }

    private static func cleanText(_ value: String, maxLength: Int, repeatLimit: Int) -> String {
        var output = ""
        var previous: Character?
        var repeatCount = 0

        for raw in value {
            let isControl = raw.unicodeScalars.allSatisfy { CharacterSet.controlCharacters.contains($0) }
            let ch: Character = raw.isWhitespace || isControl ? " " : raw
            if ch == previous {
                repeatCount += 1
                if repeatCount > repeatLimit {
                    continue
                }
            } else {
                previous = ch
                repeatCount = 1
            }

            if ch == " ", output.last == " " {
                continue
            }

            output.append(ch)
            if output.count >= maxLength {
                break
            }
        }

        return output.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
